src/Reader.cpp

```diff
@@ -57,42 +59,62 @@ RObject Reader::readToDataFrame(int lines) {
   out.attr("row.names") = IntegerVector::create(NA_INTEGER, -(rows + 1));
   out.attr("names") = outNames_;
 
-  return warnings_.addAsAttribute(out);
+  out = warnings_.addAsAttribute(out);
+
+  //collectorsClear();
+  warnings_.clear();
+
+  return out;
 }
 
 int Reader::read(int lines) {
 
+  if (t_.type() == TOKEN_EOF) {
+    return(-1);
+  }
+
   int n = (lines < 0) ? 1000 : lines;
 
   collectorsResize(n);
 
   int last_row = -1, last_col = -1, cells = 0;
-  for (Token t = tokenizer_->nextToken(); t.type() != TOKEN_EOF; t = tokenizer_->nextToken()) {
+  int first_row;
+  if (!begun_) {
+    t_ = tokenizer_->nextToken();
+    begun_ = true;
+    first_row = 0;
+  } else {
+    first_row = t_.row();
+  }
+
+  while (t_.type() != TOKEN_EOF) {
+
     if (progress_ && (++cells) % progressStep_ == 0) {
       progressBar_.show(tokenizer_->progress());
     }
 
-    if (t.col() == 0 && last_row > -1) {
+    if (t_.col() == 0 && t_.row() != first_row) {
       checkColumns(last_row, last_col, collectors_.size());
     }
 
-    if (lines >= 0 && t.row() >= lines) {
+    if (lines >= 0 && t_.row() - first_row >= lines) {
       break;
     }
 
-    if (t.row() >= n) {
+    if ((last_row - first_row) > n) {
```

Comment:
last_row will always be -1 in the first iteration, so it is ok, but still worth adding `last_row != -1` to be explicit.